File: java-server/src/main/java/com/chatsphere/server/controller/AuthController.java
package com.chatsphere.server.controller;

import com.chatsphere.server.entity.User;
import com.chatsphere.server.repository.UserRepository;
import com.chatsphere.server.security.JwtTokenProvider;
import com.chatsphere.server.service.EmailService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;
    private final EmailService emailService;

    // In-memory OTP store
    private static class OtpData {
        String otp;
        long expiresAt;
        OtpData(String otp, long expiresAt) {
            this.otp = otp;
            this.expiresAt = expiresAt;
        }
    }

    private final Map<String, OtpData> otpStore = new ConcurrentHashMap<>();
    private final Map<String, OtpData> resetOtpStore = new ConcurrentHashMap<>();

    public AuthController(UserRepository userRepository, PasswordEncoder passwordEncoder,
                          JwtTokenProvider tokenProvider, EmailService emailService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.tokenProvider = tokenProvider;
        this.emailService = emailService;
    }

    private String generateOTP() {
        return String.valueOf(100000 + new Random().nextInt(900000));
    }

    // --- 1. SEND OTP ---
    @PostMapping("/send-otp")
    public ResponseEntity<?> sendOtp(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        if (email == null || email.trim().isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Email is required"));
        }

        if (userRepository.findByEmail(email).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Email is already registered"));
        }

        String otp = generateOTP();
        long expiresAt = System.currentTimeMillis() + 5 * 60 * 1000; // 5 minutes
        otpStore.put(email, new OtpData(otp, expiresAt));

        boolean sent = emailService.sendOTPEmail(email, otp);
        if (sent) {
            return ResponseEntity.ok(Map.of("message", "OTP sent successfully"));
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to send OTP email"));
        }
    }

    // --- 2. VERIFY OTP & REGISTER ---
    @PostMapping("/verify-otp")
    public ResponseEntity<?> verifyOtp(@RequestBody Map<String, String> request) {
        String username = request.get("username");
        String email = request.get("email");
        String password = request.get("password");
        String otp = request.get("otp");

        OtpData stored = otpStore.get(email);
        if (stored == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "No OTP found. Please request a new one."));
        }

        if (!stored.otp.equals(otp)) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid Verification Code"));
        }

        if (System.currentTimeMillis() > stored.expiresAt) {
            otpStore.remove(email);
            return ResponseEntity.badRequest().body(Map.of("message", "Code has expired. Please request a new one."));
        }

        otpStore.remove(email);

        if (userRepository.findByUsername(username).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Username is already taken"));
        }

        User user = new User();
        user.setUsername(username);
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(password));
        User savedUser = userRepository.save(user);

        String token = tokenProvider.generateToken(savedUser.getId(), savedUser.getUsername());

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
            "message", "User registered successfully",
            "token", token,
            "username", savedUser.getUsername(),
            "id", savedUser.getId(),
            "profile_pic", ""
        ));
    }

    // --- 3. FORGOT PASSWORD ---
    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        if (email == null || email.trim().isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Email is required"));
        }

        if (userRepository.findByEmail(email).isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", "No account found with that email"));
        }

        String otp = generateOTP();
        long expiresAt = System.currentTimeMillis() + 5 * 60 * 1000;
        resetOtpStore.put(email, new OtpData(otp, expiresAt));

        boolean sent = emailService.sendOTPEmail(email, otp);
        if (sent) {
            return ResponseEntity.ok(Map.of("message", "Password reset OTP sent successfully"));
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Failed to send OTP email"));
        }
    }

    // --- 4. RESET PASSWORD ---
    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        String otp = request.get("otp");
        String newPassword = request.get("newPassword");

        OtpData stored = resetOtpStore.get(email);
        if (stored == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "No reset request found or it has expired."));
        }

        if (!stored.otp.equals(otp)) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid verification code"));
        }

        if (System.currentTimeMillis() > stored.expiresAt) {
            resetOtpStore.remove(email);
            return ResponseEntity.badRequest().body(Map.of("message", "Code has expired. Please request a new one."));
        }

        resetOtpStore.remove(email);

        User user = userRepository.findByEmail(email)
                .orElse(null);
        if (user == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "User not found"));
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);

        String token = tokenProvider.generateToken(user.getId(), user.getUsername());

        return ResponseEntity.ok(Map.of(
            "message", "Password reset successful",
            "token", token,
            "username", user.getUsername(),
            "id", user.getId(),
            "profile_pic", user.getProfilePic() == null ? "" : user.getProfilePic()
        ));
    }

    // --- 5. LOGIN ---
    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        String password = request.get("password");

        User user = userRepository.findByEmail(email).orElse(null);
        if (user == null || !passwordEncoder.matches(password, user.getPassword())) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("message", "Invalid email or password"));
        }

        String token = tokenProvider.generateToken(user.getId(), user.getUsername());

        return ResponseEntity.ok(Map.of(
            "message", "Login successful",
            "token", token,
            "username", user.getUsername(),
            "id", user.getId(),
            "profile_pic", user.getProfilePic() == null ? "" : user.getProfilePic()
        ));
    }

    // --- 6. REGISTER (Direct Fallback) ---
    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, String> request) {
        String username = request.get("username");
        String email = request.get("email");
        String password = request.get("password");

        if (userRepository.findByEmail(email).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Email is already registered"));
        }
        if (userRepository.findByUsername(username).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Username is already taken"));
        }

        User user = new User();
        user.setUsername(username);
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(password));
        User savedUser = userRepository.save(user);

        String token = tokenProvider.generateToken(savedUser.getId(), savedUser.getUsername());

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
            "message", "User registered successfully",
            "token", token,
            "username", savedUser.getUsername(),
            "id", savedUser.getId(),
            "profile_pic", ""
        ));
    }
}

File: server/controllers/authController.js
// Import our User helper to interact with MySQL
const User = require('../models/User');
// Import bcryptjs to hash passwords for security
const bcrypt = require('bcryptjs');
// Import jsonwebtoken to create login tokens
const jwt = require('jsonwebtoken');
// Import our new email service
const { sendOTPEmail } = require('../services/emailService');

// In-memory store for OTPs. In production, consider Redis!
// Format: { "user@example.com": { otp: "123456", expiresAt: 1234567890 } }
const otpStore = {};
const resetOtpStore = {};

// Helper function to generate a random 6-digit OTP
const generateOTP = () => {
  return Math.floor(100000 + Math.random() * 900000).toString();
};

// --- FEATURE 7: SEND OTP ---
exports.sendOtp = async (req, res) => {
  try {
    const { email } = req.body;
    
    // Check if email already exists in DB
    const existingUser = await User.findByEmail(email);
    if (existingUser) {
      return res.status(400).json({ message: 'Email is already registered' });
    }

    // Generate a 6 digit OTP
    const otp = generateOTP();
    // Set expiry to 5 minutes from now (Date.now() returns milliseconds)
    const expiresAt = Date.now() + 5 * 60 * 1000;

    // Save to our in-memory store
    otpStore[email] = { otp, expiresAt };

    // Send the email using our nodemailer service
    const emailSent = await sendOTPEmail(email, otp);
    
    if (emailSent) {
      // If email succeeded, tell frontend to move to step 2
      res.status(200).json({ message: 'OTP sent successfully' });
    } else {
      // If email failed (e.g. bad credentials), return an error
      res.status(500).json({ message: 'Failed to send OTP email. Please check server logs.' });
    }
  } catch (error) {
    console.error('Send OTP Error:', error);
    res.status(500).json({ message: 'Server error sending OTP' });
  }
};

// --- FEATURE 7: VERIFY OTP & REGISTER ---
exports.verifyOtp = async (req, res) => {
  try {
    // Frontend sends all registration details + the OTP they entered
    const { username, email, password, otp } = req.body;

    // Check if we have an OTP generated for this email
    const storedData = otpStore[email];
    if (!storedData) {
      return res.status(400).json({ message: 'No OTP found. Please request a new one.' });
    }

    // Check if OTP matches what the user entered
    if (storedData.otp !== otp) {
      return res.status(400).json({ message: 'Invalid Verification Code' });
    }

    // Check if OTP is expired
    if (Date.now() > storedData.expiresAt) {
      delete otpStore[email]; // clean up expired OTP
      return res.status(400).json({ message: 'Code has expired. Please request a new one.' });
    }

    // If we reach here, OTP is perfectly valid! 
    // Clean up the used OTP so it can't be used again
    delete otpStore[email];

    // Proceed with actual registration
    // Check username availability
    const existingUsername = await User.findByUsername(username);
    if (existingUsername) {
      return res.status(400).json({ message: 'Username is already taken' });
    }

    // Hash the password
    const hashedPassword = await bcrypt.hash(password, 10);

    // Save the new user to the MySQL database
    const newUserId = await User.createUser(username, email, hashedPassword);

    // Create a JSON Web Token (JWT) so the user is logged in automatically
    const token = jwt.sign(
      { id: newUserId, username: username },
      process.env.JWT_SECRET,
      { expiresIn: '24h' }
    );

    // Send success response with token, username, and id
    res.status(201).json({ 
      message: 'User registered successfully', 
      token,
      username: username,
      id: newUserId,
      profile_pic: null // New users have no pic yet
    });
  } catch (error) {
    console.error('Verify OTP Error:', error);
    res.status(500).json({ message: 'Server error during verification' });
  }
};

// --- FEATURE 8: FORGOT PASSWORD (SEND OTP) ---
exports.forgotPasswordOtp = async (req, res) => {
  try {
    const { email } = req.body;
    
    // Check if user exists
    const existingUser = await User.findByEmail(email);
    if (!existingUser) {
      return res.status(404).json({ message: 'No account found with that email' });
    }

    const otp = generateOTP();
    const expiresAt = Date.now() + 5 * 60 * 1000;
    resetOtpStore[email] = { otp, expiresAt };

    const emailSent = await sendOTPEmail(email, otp);
    
    if (emailSent) {
      res.status(200).json({ message: 'Password reset OTP sent successfully' });
    } else {
      res.status(500).json({ message: 'Failed to send OTP email.' });
    }
  } catch (error) {
    console.error('Forgot Password OTP Error:', error);
    res.status(500).json({ message: 'Server error sending OTP' });
  }
};

// --- FEATURE 8: RESET PASSWORD ---
exports.resetPassword = async (req, res) => {
  try {
    const { email, otp, newPassword } = req.body;

    const storedData = resetOtpStore[email];
    if (!storedData) {
      return res.status(400).json({ message: 'No reset request found or it has expired.' });
    }

    if (storedData.otp !== otp) {
      return res.status(400).json({ message: 'Invalid verification code' });
    }

    if (Date.now() > storedData.expiresAt) {
      delete resetOtpStore[email];
      return res.status(400).json({ message: 'Code has expired. Please request a new one.' });
    }

    delete resetOtpStore[email];

    // Hash the new password
    const hashedPassword = await bcrypt.hash(newPassword, 10);
    
    // Update password in DB
    await User.updatePassword(email, hashedPassword);

    // Fetch the updated user to create a token and auto-login
    const user = await User.findByEmail(email);
    
    const token = jwt.sign(
      { id: user.id, username: user.username },
      process.env.JWT_SECRET,
      { expiresIn: '24h' }
    );

    res.status(200).json({ 
      message: 'Password reset successful', 
      token,
      username: user.username,
      id: user.id,
      profile_pic: user.profile_pic
    });
  } catch (error) {
    console.error('Reset Password Error:', error);
    res.status(500).json({ message: 'Server error during password reset' });
  }
};

// --- PREVIOUS LOGIN FUNCTION ---
exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;
    const user = await User.findByEmail(email);
    
    if (!user) {
      return res.status(400).json({ message: 'Invalid email or password' });
    }

    const isMatch = await bcrypt.compare(password, user.password);
    
    if (!isMatch) {
      return res.status(400).json({ message: 'Invalid email or password' });
    }

    const token = jwt.sign(
      { id: user.id, username: user.username },
      process.env.JWT_SECRET,
      { expiresIn: '24h' }
    );

    res.status(200).json({ 
      message: 'Login successful', 
      token,
      username: user.username,
      id: user.id,
      profile_pic: user.profile_pic
    });
  } catch (error) {
    console.error('Login error:', error);
    res.status(500).json({ message: 'Server error during login' });
  }
};

// --- OLD REGISTER FUNCTION (Kept for fallback) ---
exports.register = async (req, res) => {
  try {
    const { username, email, password } = req.body;
    const existingUser = await User.findByEmail(email);
    if (existingUser) return res.status(400).json({ message: 'Email is already registered' });
    const existingUsername = await User.findByUsername(username);
    if (existingUsername) return res.status(400).json({ message: 'Username is already taken' });

    const hashedPassword = await bcrypt.hash(password, 10);
    const newUserId = await User.createUser(username, email, hashedPassword);

    const token = jwt.sign({ id: newUserId, username: username }, process.env.JWT_SECRET, { expiresIn: '24h' });
    res.status(201).json({ message: 'User registered successfully', token });
  } catch (error) {
    res.status(500).json({ message: 'Server error during registration' });
  }
};
